// rendr/server/mailer.js
var nodemailer = require('../../nodemailer');
MockmAppMailer = function() {};
MockmAppMailer.prototype.mail = function(req, callback) {
    if (typeof (req.query.EmailAddress) !== 'undefined' && req.query.EmailAddress != '') {
        MockmAppMailer.transporter = nodemailer.createTransport({
            service: 'Gmail',
            auth: {
                user: 'testprogrammer8@gmail.com',
                pass: 'testprogrammer@555'
            }
        });
        var ToEmail = [];
        if(req.query.EmailAddress instanceof Array) {
            ToEmail = req.query.EmailAddress;
        } else {
            ToEmail[0] = req.query.EmailAddress;
        }
        if(ToEmail.length > 0) {
            for(var i = 0; i < ToEmail.length; i++) {
                var MailOptions = {
                    from: 'Mockm Team<testprogrammer8@gmail.com>',
                    to: ToEmail[i].trim(),
                    subject: req.query.Subject,
                    text: '',
                    html: req.query.EmailBody
                };
                MockmAppMailer.transporter.sendMail(MailOptions, function(error, info) {
                    if (error) {
                        callback(error, '');
                    } else {
                        callback(null, info.response);
                    }
                    MockmAppMailer.transporter.close();
                });
            }
        }
    }
}
module.exports = MockmAppMailer;
